### DISCUSSION TITLE
Catch 406 errors when fetching with .single() returns no result

### DISCUSSION THREAD

**Discussion by jonohewitt** [supabase/supabase#2284] [Questions] [answered]
I'm trying to set up a function to check if a user has a username:
```
    const getUsername = async () => {
      let username: string = null
      try {
        const { data, error } = await supabase
          .from("public_user_info")
          .select("username")
          .eq("user_id", supabase.auth.user().id)
          .single()

        if (error) throw error
        if (data) username = data.username
      } catch (error) {
        // handle error
      }
      return username
    }
```

Despite this catch statement, if the user doesn't have a username yet, I get a 406 Not Acceptable error in the console:

```
details: "Results contain 0 rows, application/vnd.pgrst.object+json requires 1 row",
message: "JSON object requested, multiple (or no) rows returned"
```
Is it possible to catch this error and prevent it showing in the console? Thanks!

**Comment by inian** [comment, marked as answer]
HI @jonohewitt, I don't think the library itself prints errors to the console. Do you have a error logging setup which does this?

**Reply by hallee9000** [reply]
@inian I have the same issue. It happens when you request from an empty table.

**Reply by jonohewitt** [reply]
Sorry I missed your reply @inian ! Not that I'm aware of - I'm running the request in a Gatsby development environment so perhaps there's some error logging built in I don't know about.

**Reply by inian** [reply]
Hi @jonohewitt, I tried running this on an empty table and it doesn't print anythign in the console. I also checked https://github.com/supabase/postgrest-js for any stray console.logs and couldn't find any.

Can you provide a self contained example that reproduces the error?

```js
await supabase.from("countries").select("*").eq("id", 1).single();
```

**Reply by inian** [reply]
@leadream just to be clear, this error is expected when you call this from an empty table, but it shouldn't print anything to the console unless you explicitly print the error out. Is it printing to the console without you printing the error as well? Are you using our JS client library?

**Reply by hallee9000** [reply]
@inian Yeah, it's printed by me for debug. I can repro it.
![image](https://user-images.githubusercontent.com/12868055/127800335-e4c4ea12-aedf-4aac-bd39-8b2a3565cb72.png)

live code here
https://codesandbox.io/s/dawn-architecture-1esl8

**Reply by inian** [reply]
Hi Jun, that is expected. Since you are printing the error at line 13, it shows the error in the console.

**Reply by hallee9000** [reply]
@inian I know it's expected but this is not the point. I mean this code should not return 406, should return 200 and `null`

**Reply by inian** [reply]
This matches the postgrest behaviour that we use under the hood Jun. If are expecting a result to not have a single response, call the query without `single`. You should get an empty result then.

**Reply by hallee9000** [reply]
Get it, thanks @inian, though I still think it not intuitive.

**Reply by ManuLpz4** [reply]
Now you can just use ``maybeSingle()`` instead of ``limit(1).single()``, it allows you to fetch zero or one single object of the query. Notice that it doesn't need the limit filter.

https://supabase.com/docs/reference/javascript/maybesingle

**Reply by henningko** [reply]
However, even with maybeSingle, there will be a 406 error if there are 0 rows returned. It seems that one would want to use maybeSingle() over single() to avoid such error, both for the network request and in the JS response. In vanilla JS, this causes me to not be able to handle a 0-row response.

**Reply by devonkcopeland-serial** [reply]
I'm having the same issue as @henningko and @leadream 

I've been playing around `.single()` and `.maybeSingle()`, but I'm still getting 406 errors thrown with `.limit(1).single()` and `.maybeSingle()` when a matching row doesn't exist.  I've also tried catching the 406 with a try catch block but the errors are making their way to the console from `fetch.ts`.

**Reply by steve-chavez** [reply]
>  I've also tried catching the 406 with a try catch block but the errors are making their way to the console from fetch.ts.

This will be addressed on https://github.com/supabase/postgrest-js/issues/361

**Reply by otang** [reply]
I'm still seeing this issue.

**Reply by jacobfogolyan** [reply]
I too am seeing this, i have recently re-visited a project that was in working order and now i seem to be getting these errors in multiple places. 

EDIT: Apologies, this was a testing error it was due to RLS rules i put in place

**Reply by Psaw19** [reply]
thanks my issue was the same and it got resolved when i used maybeSingle() instead of single()

**Reply by HORACINIS** [reply]
Thanks @Psaw19 that maybeSingle() did the trick for me!

**Reply by MateoWartelle** [reply]
Thanks @Psaw19

**Reply by dadashvespek** [reply]
thanks! @Psaw19

**Reply by salimdellali** [reply]
`mayeSingle()` did the trick for me too, thank's @Psaw19 

I had this table `public.accounts` (with RLS enabled) that maps 1 to 1 with `auth.users` (`public.accounts` has a column called `user_id` which is a foreign key referencing `id` in `auth.users`). I have this state in my system where when a user signs up, supabase automatically  inserts a record into `auth.users` (but not in `public.accounts`), then the user is invited to fill in his account information, it is then where a new row is inserted into `public.accounts`

So when I do
```ts
const { data, error } = await supabase
        .from("accounts")
        .select("*")
```
the resulted `data` is either:
- an empty array `[]` (a.k.a no account associated to the authenticated user (yet)) 
- an array of 1 and only 1 account object, containing the account information associated with the authenticated user

so appending `maybeSingle()` after the `.select("*")` resulted in `data` be either:
- `null` if no account is associated to the authenticated user
- an account object with information associated with the authenticated user

**Reply by joepolask86** [reply]
> thanks my issue was the same and it got resolved when i used maybeSingle() instead of single()

This will did the job for me.
